File: app/api/src/main/kotlin/com/wallet/clover/api/service/WinningCheckService.kt
package com.wallet.clover.api.service

import com.wallet.clover.api.entity.game.LottoGameEntity
import com.wallet.clover.api.entity.game.LottoGameStatus
import com.wallet.clover.api.entity.winning.WinningInfoEntity
import com.wallet.clover.api.repository.game.LottoGameRepository
import com.wallet.clover.api.repository.ticket.LottoTicketRepository
import com.wallet.clover.api.repository.winning.WinningInfoRepository
import kotlinx.coroutines.flow.toList
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class WinningCheckService(
    private val winningInfoRepository: WinningInfoRepository,
    private val lottoTicketRepository: LottoTicketRepository,
    private val lottoGameRepository: LottoGameRepository
) {
    private val logger = LoggerFactory.getLogger(WinningCheckService::class.java)

    @Transactional
    suspend fun checkWinning(round: Int) {
        // 1. 해당 회차 당첨 정보 조회
        val winningInfo = winningInfoRepository.findByRound(round)
        if (winningInfo == null) {
            logger.warn("Winning info for round $round not found.")
            return
        }

        // 2. 해당 회차의 티켓 조회
        val tickets = lottoTicketRepository.findByOrdinal(round)
        logger.info("Found ${tickets.size} tickets for round $round")

        for (ticket in tickets) {
            val games = lottoGameRepository.findByTicketId(ticket.id!!)
            var ticketTotalPrize = 0L
            var hasWinningGame = false

            for (game in games) {
                val (statusStr, prize) = calculatePrize(game, winningInfo)
                val status = LottoGameStatus.valueOf(statusStr)
                
                if (game.status != status || game.prizeAmount != prize) {
                    val updatedGame = game.copy(
                        status = status,
                        prizeAmount = prize
                    )
                    lottoGameRepository.save(updatedGame)
                    logger.debug("Updated game ${game.id}: $status, $prize")
                }
                
                if (prize > 0) {
                    ticketTotalPrize += prize
                    hasWinningGame = true
                }
            }
            
            // 티켓 상태 업데이트 (당첨된 게임이 하나라도 있으면 WINNING, 아니면 LOSING)
            // 단, 아직 추첨 전인 경우는 처리하지 않음 (이 메서드는 추첨 후 호출됨)
            // val newTicketStatus = if (hasWinningGame) "WINNING" else "LOSING"
            // TODO: LottoTicketStatus Enum 변환 및 티켓 상태 업데이트 구현 필요
        }
    }
    
    fun calculatePrize(game: LottoGameEntity, winningInfo: WinningInfoEntity): Pair<String, Long> {
        val myNumbers = setOf(game.number1, game.number2, game.number3, game.number4, game.number5, game.number6)
        val winningNumbers = setOf(winningInfo.number1, winningInfo.number2, winningInfo.number3, winningInfo.number4, winningInfo.number5, winningInfo.number6)
        
        val matchCount = myNumbers.intersect(winningNumbers).size
        val bonusMatch = myNumbers.contains(winningInfo.bonusNumber)
        
        return when (matchCount) {
            6 -> "WINNING_1" to winningInfo.firstPrizeAmount
            5 -> if (bonusMatch) "WINNING_2" to winningInfo.secondPrizeAmount else "WINNING_3" to winningInfo.thirdPrizeAmount
            4 -> "WINNING_4" to winningInfo.fourthPrizeAmount
            3 -> "WINNING_5" to winningInfo.fifthPrizeAmount
            else -> "LOSING" to 0L
        }
    }
}

File: apps/backend/src/lotto/winning-check.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { OnEvent } from '@nestjs/event-emitter';
import { LottoRankCalculator } from './utils/lotto-rank.calculator';
import { WinnerNotificationService } from './winner-notification.service';





@Injectable()
export class WinningCheckService {
  private readonly logger = new Logger(WinningCheckService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly winnerNotification: WinnerNotificationService,
  ) {}

  @OnEvent('lotto.winning-info.created')
  async handleWinningInfoCreated(payload: { round: number }) {
    this.logger.log(`${payload.round} 회차 당첨 정보 생성 이벤트 수신`);
    await this.checkWinning(payload.round);
  }

  async checkWinning(round: number) {
    const winningInfo = await this.prisma.winningInfo.findUnique({
      where: { round },
    });

    if (!winningInfo) {
      this.logger.warn(`${round} 회차 당첨 정보를 찾을 수 없습니다.`);
      return;
    }

    const tickets = await this.prisma.lottoTicket.findMany({
      where: { ordinal: round },
      include: { games: true },
    });

    this.logger.log(`${round} 회차 티켓 ${tickets.length}건 처리 시작`);

    for (const ticket of tickets) {
      await this.processTicket(ticket, winningInfo);
    }
  }

  private async processTicket(ticket: any, winningInfo: any) {
    try {
      const { hasWinningGame, updates } = this.calculateTicketResults(
        ticket,
        winningInfo,
      );

      if (updates.length > 0) {
        await Promise.all(updates.map((u) => this.prisma.lottoGame.update(u)));
      }

      const newStatus = hasWinningGame ? 'WINNING' : 'LOSING';
      if (ticket.status !== newStatus) {
        await this.updateTicketStatus(ticket, newStatus, winningInfo);
      }
    } catch (error) {
      this.logger.error(`티켓 ${ticket.id} 처리 중 오류 발생`, error.stack);
    }
  }

  private calculateTicketResults(ticket: any, winningInfo: any) {
    let hasWinningGame = false;
    const updates = [];

    const prizeAmounts = {
      WINNING_1: winningInfo.firstPrizeAmount,
      WINNING_2: winningInfo.secondPrizeAmount,
      WINNING_3: winningInfo.thirdPrizeAmount,
      WINNING_4: winningInfo.fourthPrizeAmount,
      WINNING_5: winningInfo.fifthPrizeAmount,
    };

    for (const game of ticket.games) {
      const gameNumbers = [
        game.number1,
        game.number2,
        game.number3,
        game.number4,
        game.number5,
        game.number6,
      ];
      const winNumbers = [
        winningInfo.number1,
        winningInfo.number2,
        winningInfo.number3,
        winningInfo.number4,
        winningInfo.number5,
        winningInfo.number6,
      ];

      const { status, prize } = LottoRankCalculator.calculateRank(
        gameNumbers,
        winNumbers,
        winningInfo.bonusNumber,
        prizeAmounts,
      );

      if (game.status !== status || game.prizeAmount !== prize) {
        updates.push({
          where: { id: game.id },
          data: { status, prizeAmount: prize },
        });
      }

      if (prize > BigInt(0)) hasWinningGame = true;
    }

    return { hasWinningGame, updates };
  }

  private async updateTicketStatus(
    ticket: any,
    status: string,
    winningInfo: any,
  ) {
    await this.prisma.lottoTicket.update({
      where: { id: ticket.id },
      data: { status },
    });

    if (status === 'WINNING') {
      this.winnerNotification.handleWinnerActions(ticket, winningInfo);
    }
  }
}
